technic/machines/network.lua

```diff
@@ -0,0 +1,587 @@
+--
+-- Power network specific functions and data should live here
+--
+local S = technic.getter
+
+local switch_max_range = tonumber(minetest.settings:get("technic.switch_max_range") or "256")
+
+technic.networks = {}
+technic.cables = {}
+
+local poshash = minetest.hash_node_position
+local hashpos = minetest.get_position_from_hash
+
+function technic.create_network(sw_pos)
+	local network_id = poshash({x=sw_pos.x,y=sw_pos.y-1,z=sw_pos.z})
+	technic.build_network(network_id)
+	return network_id
+end
+
+function technic.activate_network(network_id, timeout)
+	assert(network_id)
+	local network = technic.networks[network_id]
+	if network and (timeout or network.timeout < 4) then
+		network.timeout = timeout or 4
+	end
+end
+
+function technic.sw_pos2tier(pos, use_vm)
+	-- Get cable tier for switching station or nil if no cable
+	-- use_vm true to use VoxelManip to load node
+	local cable_pos = {x=pos.x,y=pos.y-1,z=pos.z}
+	if use_vm then
+		technic.get_or_load_node(cable_pos)
+	end
+	return technic.get_cable_tier(minetest.get_node(cable_pos).name)
+end
+
+-- Destroy network data
+function technic.remove_network(network_id)
+	local cables = technic.cables
+	for pos_hash,cable_net_id in pairs(cables) do
+		if cable_net_id == network_id then
+			cables[pos_hash] = nil
+		end
+	end
+	technic.networks[network_id] = nil
+	--print(string.format("technic.remove_network(%.17g) at %s", network_id, minetest.pos_to_string(technic.network2pos(network_id))))
+end
+
+-- Remove machine or cable from network
+local network_node_arrays = {"PR_nodes","BA_nodes","RE_nodes","SP_nodes"}
+function technic.remove_network_node(network_id, pos)
+	local network = technic.networks[network_id]
+	if not network then return end
+	-- Clear hash tables, cannot use table.remove
+	local node_id = poshash(pos)
+	technic.cables[node_id] = nil
+	network.all_nodes[node_id] = nil
+	-- Clear indexed arrays, do NOT leave holes
+	for _,tblname in ipairs(network_node_arrays) do
+		local table = network[tblname]
```

Comment:
You should probably use some other name here :wink:

Reply:
probably, did not change names when I copied that loop from register/cables.lua
however my imagination is not any better than op's :smile:

Reply:
> Experienced two crashes with [0d0a0bb](https://github.com/mt-mods/technic/commit/0d0a0bbbb428563117fc748ce4eabbd2252c01f8):
> 
> While a player was connecting two networks (built before switching to this branch) with some machines but no switching station (I think):

Almost sure bug that caused crash was introduced in last significant commit which was dd8fcb4
That was not so well tested, especially digging not tested at all.
Also after breaking register/cables.lua apart I had similar crashes with placement but fixed most of those if not all, digging part was left without much love and apparently needs some work.

I'll see if I can reproduce it and possibly add unit test for situation.